src/core/layout/qgscompositionconverter.cpp

```diff
@@ -0,0 +1,1414 @@
+}
+
+
+QgsLayout *QgsCompositionConverter::createLayoutFromCompositionXml( const QDomElement &parentElement, QgsProject *project )
+{
+  initPropertyDefinitions();
+  QgsLayout *layout = new QgsLayout( project );
+  // Create pages
+  int pages = parentElement.attribute( QStringLiteral( "numPages" ) ).toInt( );
+  float paperHeight = parentElement.attribute( QStringLiteral( "paperHeight" ) ).toFloat( );
+  float paperWidth = parentElement.attribute( QStringLiteral( "paperWidth" ) ).toFloat( );
+
+  if ( parentElement.elementsByTagName( QStringLiteral( "symbol" ) ).size() )
+  {
+    QDomElement symbolElement = parentElement.elementsByTagName( QStringLiteral( "symbol" ) ).at( 0 ).toElement();
+    QgsReadWriteContext context;
+    context.setPathResolver( project->pathResolver() );
+    QgsFillSymbol *symbol = QgsSymbolLayerUtils::loadSymbol<QgsFillSymbol>( symbolElement, context );
+    if ( symbol )
+      layout->pageCollection()->setPageStyleSymbol( symbol );
+  }
+
+  QString name = parentElement.attribute( QStringLiteral( "name" ) );
+  layout->setName( name );
+  // TODO: check that it is always landscape
+  QgsLayoutSize pageSize( paperWidth, paperHeight );
+  for ( int j = 0; j < pages; j++ )
+  {
+    QgsLayoutItemPage *page = QgsLayoutItemPage::create( layout );
+    page->setPageSize( pageSize );
+    layout->pageCollection()->addPage( page );
+  }
+  addItemsFromCompositionXml( layout, parentElement );
+  return layout;
+}
+
+void QgsCompositionConverter::adjustPos( QgsLayout *layout, QgsLayoutItem *layoutItem, QDomNode &itemNode, QPointF *position, bool &pasteInPlace, int zOrderOffset, QPointF &pasteShiftPos, int &pageNumber )
+{
+  if ( position )
+  {
+    if ( pasteInPlace )
+    {
+      QgsLayoutPoint posOnPage = QgsLayoutPoint::decodePoint( itemNode.toElement().attribute( QStringLiteral( "positionOnPage" ) ) );
+      layoutItem->attemptMove( posOnPage, true, false, pageNumber );
+    }
+    else
+    {
+      layoutItem->attemptMoveBy( pasteShiftPos.x(), pasteShiftPos.y() );
+    }
+  }
+
+  layout->addLayoutItem( layoutItem );
+  layoutItem->setZValue( layoutItem->zValue() + zOrderOffset );
+}
+
+void QgsCompositionConverter::restoreGeneralComposeItemProperties( QgsLayoutItem *layoutItem, const QDomElement &itemElem )
+{
+  //restore general composer item properties
+  QDomNodeList composerItemList = itemElem.elementsByTagName( QStringLiteral( "ComposerItem" ) );
+  if ( !composerItemList.isEmpty() )
+  {
+    QDomElement composerItemElem = composerItemList.at( 0 ).toElement();
+
+    //rotation
+    if ( !qgsDoubleNear( composerItemElem.attribute( QStringLiteral( "rotation" ), QStringLiteral( "0" ) ).toDouble(), 0.0 ) )
+    {
+      //check for old (pre 2.1) rotation attribute
+      layoutItem->setItemRotation( composerItemElem.attribute( QStringLiteral( "rotation" ), QStringLiteral( "0" ) ).toDouble(), false );
+    }
+
+    QgsCompositionConverter::readXml( layoutItem, composerItemElem );
+
+    // Frame color
```

Comment:
Are these a TODO?

Reply:
Yes, I took the enums and used them as TODOs, I can remove it.